docs/design/map/map-store-operation-offloading.md

```diff
@@ -0,0 +1,216 @@
+    <write-coalescing>true</write-coalescing>
+</map-store>
+```
+```yaml
+map-store:
+  enabled: true
+  initial-mode: LAZY
+  class-name: com.hazelcast.examples.DummyStore
+  offload: true
+  write-delay-seconds: 60
+  write-batch-size: 1000
+  write-coalescing: true
+```
+##### Configuring Offloadable Executor
+`hz:map-store-offloadable` is the name of executor
+which `Step`s are offloaded on. It can be configured
+like regular Hazelcast executors via `ExecutorConfig`
+
+#### Implementation
+We divide MapStore interaction needed operations into parts which can be executed separately.
+These self executable parts are called `Step`:
+```java
+interface Step {
+
+    void runStep(State state);
+
+    Step nextStep(State state);
+
+    boolean isOffloadStep();
+}
+```
+In its essence, a step is either an instance of a `PartitionSpecificRunnable` or a `Runnable`.
+
+And we model an operation as a sequence of `Step`s.
+
+#### Example Model for IMap#get
+For instance, sequence of get operation steps are: `READ`,`LOAD`,`ON_LOAD`,`RESPONSE`,`AFTER_RUN`. 
+
+`READ` step tries to read value in partition thread and if it finds no matching
+value in memory, `LOAD` step starts value loading from MapStore in a separate offloaded thread. 
+Upon `LOAD` finishes, execution is restarted in partition thread again from the step `ON_LOAD`.
+
+If no loading from MapStore is required, all steps are
+executed in partition thread and no offloading is done.
+
+`State` is used to pass state between `Step`s.
+
+#### Improvement for Ordering of Operations
+In previous Hazelcast versions, all map operations wait end of
+blocking MapStore operation than they progress. This provides an
+order of execution between operations. Not to break this behavior,
+in this new implementation, there are offloaded operation queues,
+these queues are per map per partition queues(so a map has a separate
+queue inside a partition) and all map operations that require MapStore
+interaction waits in offloaded queue if there is an in-flight operation
+for that specific map partition. Upon finish of the in-flight
+operation, next operation is taken from the queue and starts to execute.
+
+This change is an improvement since only a specific map's
+operations will be waiting each other and partitions are
+free for other processing. In previous versions, all maps are
+waiting one single blocking map and partitions are blocked. 
+
+#### Queued Offloaded Operations During Migration
+Operations on source side of migration are retried on migration commit.
+
+
+#### Flow in Partition Thread
+
+                      ------------------------------
+                      Operation starts its execution
+                      ------------------------------
+                                  |
+                                  |
+          ------------------------------------------------------
+            Check if the operation needs mapstore interaction?  ---> If No --> Do regular execution
```

Comment:
We now have 2 implementations for map operations that are offloadable - the original which is used when offload is false, or when there is no map store, and the new implementation which is used when we have a map store and offload set to true (default).

- changes to the operations will need to be done in 2 places, this will inevitably lead to bugs when we forget to update one place or another.
- the offload implementation is not tested as much as the one without offload because most tests don't use mapstore, this makes it rather risky.

I see this as the main issue with the PR.